### PR TITLE
WIP: API Overview udpates

### PR DESCRIPTION
**Pull request by pavanpodila**
Handling the updates to the **API Overview** page.

### CODE REVIEW

File: docs/refguide/api.md [hunk 1]
```diff
@@ -1,10 +1,16 @@
 # MobX Api Reference
 
-Applies to MobX 3 and higher. For MobX 2, the old documentation is still available on [github](https://github.com/mobxjs/mobx/blob/7c9e7c86e0c6ead141bb0539d33143d0e1f576dd/docs/refguide/api.md).
+**Applies to MobX 4 and higher**
+
+- Using Mobx 3? Use this [migration guide](https://github.com/mobxjs/mobx/wiki/Migrating-from-mobx-3-to-mobx-4) to switch gears.
+- For MobX 2, the old documentation is still available on [github](https://github.com/mobxjs/mobx/blob/7c9e7c86e0c6ead141bb0539d33143d0e1f576dd/docs/refguide/api.md).
```
Comment (mweststrate) [inline docs/refguide/api.md]: Also: MobX 3 documentation link: [MobX 3](https://github.com/mobxjs/mobx/blob/54557dc319b04e92e31cb87427bef194ec1c549c/docs/refguide/api.md)

File: docs/refguide/api.md [hunk 2]
```diff
@@ -15,21 +21,26 @@ Usage:
...
 1. If *value* is an array, a new [Observable Array](array.md) will be returned.
 1. If *value* is an object *without* prototype or its prototype is `Object.prototype`, the object will be cloned and all its current properties will be made observable. See [Observable Object](object.md)
-1. If *value* is an object *with* a prototype, a JavaScript primitive or function, a [Boxed Observable](boxed.md) will be returned. MobX will not make objects with a prototype automatically observable; as that is the responsibility of its constructor function. Use `extendObservable` in the constructor, or `@observable` in its class definition instead.
+1. If *value* is an object *with* a prototype, a JavaScript primitive or function, there will be no change made to the value. If you do need a [Boxed Observable](boxed.md), call `observable.box(value)` explicitly. MobX will not make objects with a prototype automatically observable; as that is the responsibility of its constructor function. Use `extendObservable` in the constructor, or `@observable` in its class definition instead.
```
Comment (mweststrate) [inline docs/refguide/api.md]: ... use `@observable` in the class definition, or alternatively `[decorate](link)` or  `[extendObservable](link)` to introduce observable properties on a class definition.

File: docs/refguide/api.md [hunk 3]
```diff
@@ -40,72 +51,88 @@ sugar for `extendObservable(this, { property: value })`.
 
 [&laquo;`details`&raquo;](observable-decorator.md)
 
-### `observable.box(value)` & `observable.shallowBox(value)`
-
+### `observable.box(value)` & `observable.box(value, {deep: false})`
```
Comment (mweststrate) [inline docs/refguide/api.md]: maybe just `observable.box(value, options?)` also holds for the items below

File: docs/refguide/api.md [hunk 4]
```diff
@@ -40,72 +51,88 @@ sugar for `extendObservable(this, { property: value })`.
...
 }
 ```
 
-Or as property modifiers in combination with `observable.object` / `observable.extendObservable`.
-Note that modifiers always 'stick' to the property. So they will remain in effect even if a new value is assigned.
+Or by passing in property decorators via `observable.object` / `observable.extendObservable`.
+Note that decorators always 'stick' to the property. So they will remain in effect even if a new value is assigned.
 
 ```javascript
+import {observable, action} from 'mobx';
```
Comment (mweststrate) [inline docs/refguide/api.md]: also mention `decorate` in this section?

File: docs/refguide/api.md [hunk 5]
```diff
@@ -119,14 +146,19 @@ Usage:
 * `computed(() => expression, (newValue) => void)`
 * `computed(() => expression, options)`
 * `@computed get classProperty() { return expression; }`
-* `@computed.struct get classProperty() { return expression; }`
```
Comment (mweststrate) [inline docs/refguide/api.md]: `.struct` still exists

File: docs/refguide/api.md [hunk 6]
```diff
@@ -138,18 +170,18 @@ With MobX you can make it explicit in your code where your actions live by marki
...
-Note: using `action` is mandatory when *strict mode* is enabled, see `useStrict`.
+Note: using `action` is mandatory when *strict mode* is enabled, see `enforceActions`.
 [&laquo;details&raquo;](action.md)
 
 Usage:
 * `action(fn)`
 * `action(name, fn)`
 * `@action classMethod`
 * `@action(name) classMethod`
-* `@action boundClassMethod = (args) => { body }`
```
Comment (mweststrate) [inline docs/refguide/api.md]: These first two are still valid

File: docs/refguide/api.md [hunk 7]
```diff
@@ -138,18 +170,18 @@ With MobX you can make it explicit in your code where your actions live by marki
...
-* `@action(name) boundClassMethod = (args) => { body }`
+* `@action.bound boundClassMethod = (args) => { body }`
+* `@action.bound(name) boundClassMethod = (args) => { body }`
```
Comment (mweststrate) [inline docs/refguide/api.md]: `@action.bound(name)` doesn't exist.

However, `@action.bound method(args) { body }` does work

File: docs/refguide/api.md [hunk 8]
```diff
@@ -172,36 +204,38 @@ Usage:
...
 
 ### `autorun`
-Usage: `autorun(debugname?, () => { sideEffect })`. Autorun runs the provided `sideEffect` and tracks which observable state is accessed while running the side effect.
+Usage: `autorun(() => { sideEffect }, options)`. Autorun runs the provided `sideEffect` and tracks which observable state is accessed while running the side effect.
 Whenever one of the used observables is changed in the future, the same sideEffect will be run again.
 Returns a disposer function to cancel the side effect. [&laquo;details&raquo;](autorun.md)
 
+**options**
+- **`debugname?: string`**: A name for easier identification and debugging
+- **`delay?: number`**: the sideEffect will be delayed and debounced with the given `delay`. Defaults to `0`.
```
Comment (mweststrate) [inline docs/refguide/api.md]:
```
- ** `onError?: (error) => void` **: error handler that will be triggered if the autorun function throws an exception
- ** `scheduler?: (callback) => void` **: Set a custom scheduler to determine how re-running the autorun function should be scheduled
```

Comment (pavanpodila) [reply docs/refguide/api.md]: I was thinking if keeping the advanced options in the details doc. But probably makes sense to include here too

File: docs/refguide/api.md [hunk 9]
```diff
@@ -172,36 +204,38 @@ Usage:
...
+
 ### `when`
-Usage: `when(debugname?, () => condition, () => { sideEffect })`.
+Usage: `when(() => condition, () => { sideEffect }, options)`.
```
Comment (mweststrate) [inline docs/refguide/api.md]: effect is now optional, if no effect is specified, returns a (cancellable) promise

options:
- `name`
- `timeout`
- `onError`

File: docs/refguide/api.md [hunk 10]
```diff
@@ -172,36 +204,38 @@ Usage:
...
 
-### `expr`
-Usage: `expr(() => someExpression)`. Just a shorthand for `computed(() => someExpression).get()`.
-`expr` is useful in some rare cases to optimize another computed function or reaction.
-In general it is simpler and better to just split the function in multiple smaller computed's to achieve the same effect.
-[&laquo;details&raquo;](expr.md)
+options
+**options**
+- **`fireImmediately?: boolean`**: Wait for a change before firing the _effect function_. Defaults to `false`.
+- **`delay?: number`**: the sideEffect will be delayed and debounced with the given `delay`. Defaults to `0`.
```
Comment (mweststrate) [inline docs/refguide/api.md]: options: same ones as autorun, + `fireImmediately` to immediately run the effect when the reaction is created (`false` by default)

File: docs/refguide/api.md [hunk 11]
```diff
@@ -228,8 +262,8 @@ Higher order component and counterpart of `Provider`. Can be used to pick stores
 ### `toJS`
 Usage: `toJS(observableDataStructure)`. Converts observable data structures back to plain javascript objects, ignoring computed values. [&laquo;details&raquo;](tojson.md)
```
Comment (mweststrate) [inline docs/refguide/api.md]: `toJS` accepts options now as well:

```
 {
    detectCycles: true,
    exportMapsAsObjects: true
}
```

Comment (mweststrate) [reply docs/refguide/api.md]: (both true by default)

File: docs/refguide/api.md [hunk 12]
```diff
@@ -228,8 +262,8 @@ Higher order component and counterpart of `Provider`. Can be used to pick stores
...
 
-### `isObservable`
```
Comment (mweststrate) [inline docs/refguide/api.md]: `isObservable` still exists, but has only one argument

File: docs/refguide/api.md [hunk 13]
```diff
@@ -243,14 +277,8 @@ Note that observable arrays can be `.slice()`d to turn them into true JS-arrays.
...
 
-### `isComputed`
-Usage: `isComputed(thing, property?)`. Returns true if the given thing is a boxed computed value, or if the designated property is a computed value.
-
-### `createTransformer`
-Usage: `createTransformer(transformation: A => B, onCleanup?): A = B`.
-Can be used to make functions that transforms one value into another value reactive and memoized.
-It behaves similar to computed and can be used for advanced patterns like very efficient array maps, map reduce or computed values that are not part of an object.
-[&laquo;details&raquo;](create-transformer.md)
+### `isComputedProp`
```
Comment (mweststrate) [inline docs/refguide/api.md]: `isComputed` still exists as well, but with one argument

File: docs/refguide/api.md [hunk 14]
```diff
@@ -262,12 +290,57 @@ Usage: `observe(object, property?, listener, fireImmediately = false)`
 Low-level api that can be used to observe a single observable value.
 [&laquo;details&raquo;](observe.md)
 
-### `useStrict`
-Usage: `useStrict(boolean)`.
-Enables / disables strict mode *globally*.
+### `decorate`
```
Comment (mweststrate) [inline docs/refguide/api.md]: probably this one should be moved to the observable section in the beginning?

File: docs/refguide/api.md [hunk 15]
```diff
@@ -138,18 +170,18 @@ With MobX you can make it explicit in your code where your actions live by marki
...
 * `action(name, fn)`
 * `@action classMethod`
 * `@action(name) classMethod`
-* `@action boundClassMethod = (args) => { body }`
-* `@action(name) boundClassMethod = (args) => { body }`
+* `@action.bound boundClassMethod = (args) => { body }`
+* `@action.bound(name) boundClassMethod = (args) => { body }`
 
-For one-time-actions `runInAction(name?, fn, scope?)` can be used, which is sugar for `action(name, fn, scope)()`.
+For one-time-actions `runInAction(name?, fn)` can be used, which is sugar for `action(name, fn)()`.
```
Comment (mweststrate) [inline docs/refguide/api.md]: add `flow` below `action`

**Comment by mweststrate** [comment]
I think it is nice to mention them here for now. At some I want to
restructure the docs a bit, with coarsly the structure of this document,
but with more narrative & examples, instead of all the per-api pages we now
have.

Op za 10 mrt. 2018 om 11:08 schreef Pavan Podila <notifications@github.com>:

File: docs/refguide/api.md (hunk 15 above)
Comment (pavanpodila) [reply docs/refguide/api.md]: By `flow` you mean a type signature ?

File: docs/refguide/api.md (hunk 10 above)
Comment (pavanpodila) [reply docs/refguide/api.md]: added

File: docs/refguide/api.md (hunk 11 above)
Comment (pavanpodila) [reply docs/refguide/api.md]: added

File: docs/refguide/api.md (hunk 12 above)
Comment (pavanpodila) [reply docs/refguide/api.md]: added back

File: docs/refguide/api.md (hunk 14 above)
Comment (pavanpodila) [reply docs/refguide/api.md]: yup moved just under decorators

**Comment by pavanpodila** [comment]
Added section on flow too ;-)

**Comment by mweststrate** [comment]
Thanks!

**Pull request merged**